pkg/virt-launcher/notify-client/client.go

```diff
@@ -247,38 +247,50 @@ func eventCallback(c cli.Connection, domain *api.Domain, libvirtEvent libvirtEve
 		domain.ObjectMeta.DeletionTimestamp = &now
 		watchEvent := watch.Event{Type: watch.Modified, Object: domain}
 		client.SendDomainEvent(watchEvent)
-		events <- watchEvent
+		updateEvents(watchEvent, domain, events)
 	default:
 		if libvirtEvent.Event != nil {
 			if libvirtEvent.Event.Event == libvirt.DOMAIN_EVENT_DEFINED && libvirt.DomainEventDefinedDetailType(libvirtEvent.Event.Detail) == libvirt.DOMAIN_EVENT_DEFINED_ADDED {
 				event := watch.Event{Type: watch.Added, Object: domain}
 				client.SendDomainEvent(event)
-				events <- event
+				updateEvents(event, domain, events)
 			} else if libvirtEvent.Event.Event == libvirt.DOMAIN_EVENT_STARTED && libvirt.DomainEventStartedDetailType(libvirtEvent.Event.Detail) == libvirt.DOMAIN_EVENT_STARTED_MIGRATED {
 				event := watch.Event{Type: watch.Added, Object: domain}
 				client.SendDomainEvent(event)
-				events <- event
+				updateEvents(event, domain, events)
 			}
 		}
-		log.Log.Infof("7) OSINFO IN EVENT CALLBACK: %v", osInfo)
 		if interfaceStatus != nil {
 			domain.Status.Interfaces = interfaceStatus
 		}
 		if osInfo != nil {
 			domain.Status.OSInfo = *osInfo
 		}
-		if interfaceStatus != nil || osInfo != nil {
-			event := watch.Event{Type: watch.Modified, Object: domain}
-			client.SendDomainEvent(event)
-			events <- event
-		}
+
 		err := client.SendDomainEvent(watch.Event{Type: watch.Modified, Object: domain})
 		if err != nil {
 			log.Log.Reason(err).Error("Could not send domain notify event.")
 		}
 	}
 }
 
+var updateEvents = updateEventsClosure()
+
+func updateEventsClosure() func(event watch.Event, domain *api.Domain, events chan watch.Event) {
+	firstAddEvent := true
+	firstDeleteEvent := true
```

Comment:
i've never seen this pattern before. Are firstAddEvent and firstDeleteEvent global somehow? Also, is this thread safe?

Reply:
i see your github comment now. Sorry i missed that. I'm reading up on this now.

Reply:
its not thread safe, but since `eventCallback` is running in serial only (by its caller), its safe

Reply:
i'm pretty sure this doesn't work.


```
package main

import "fmt"

func main() {
	y := func() {
		x := 1
		x++
		fmt.Println("x:", x)
		
	}

	for i := 0; i < 10; i++ {
		y()
	}
}
```

always prints 2.

Reply:
you must have x outside of the function
and then it will work
i made sure with prints, and without it i saw the channel get full, with it, it doesnt

Reply:
sure, that example works. but that's not what the logic  in this PR. the variables are defined inside the closure in this PR.

Reply:
its the same, thing, there are 2 functions there, this is why there is layering here
https://github.com/kubevirt/kubevirt/pull/4253/files#diff-7175e08324a35a860b12d6518d396988R277

i will create a little playground example

Reply:
https://play.golang.org/p/b84xdWx5vq5

Reply:
oh, i get it now. you're returning a function. that's what makes it work.